rfcs/20220925-bounded-dynamism.md

```diff
@@ -0,0 +1,255 @@
+# StableHLO Bounded Dynamism
+
+[StableHLO](https://github.com/openxla/stablehlo) is an operation set that
+expresses ML computations. It has been originally bootstrapped from the [MHLO
+dialect](https://github.com/tensorflow/mlir-hlo#meta-hlo-dialect-mhlo),
+including inheriting the type and some ops. This RFC aims to describe the
+current status and rationale for bounded dynamism constructs in StableHLO.  In
+particular, this RFC doesn’t propose any further changes to the current state.
+
+Bounded dynamism allows programs to represent the maximum runtime size that a
+particular dynamic dimension of a tensor can have. This makes it possible to run
+such programs on hardware that doesn't support dynamic tensors but could support
+it if the upper bounds of tensor dimensions are known at the compilation time.
+Applications include:
+
+* Real time inference without having to wait for accumulation up to a particular
+  batch size.
+* Programs whose intermediate tensor shapes depend on the operands. For example,
+  [`stablehlo.dynamic_broadcast_in_dim`](https://github.com/openxla/stablehlo/blob/ff55f9346d54e9e38de807a79f8ae03faffda274/stablehlo/dialect/StablehloOps.td#L1838)
+  op but with statically known upper bounds of `output_dimensions` operand.
+* Bounded dynamism can also open up performance optimizations opportunities.
+
+## Non Goals
+
+* Provide [value inference](https://github.com/openxla/xla/blob/9e05932a2ceadea080dc9494cfe9d735f94c4e68/xla/client/value_inference.h)
+  like utility for producers that want to generate the `set_dimension_size` ops.
+  Value inference depends on constant folding for StableHLO ops which is a work
+  in progress currently. There will be separate RFC for value inference subject
+  to separate approval. Note that producers that generate unbounded programs
+  don't need this
+  in StableHLO.
+* Provide a transformation that converts StableHLO programs to bounded StableHLO
+  programs. However, there is a plan to have such a conversion in MHLO, although
+  the details of this are out of scope of this RFC.
+
+## Detailed Proposal
+
+### (P1) Bounded tensor type using the encoding field in the RankedTensorType
```

Comment:
Lets start with adding a "wrapper" class so that one can use LLVM style RTTI (e.g., dyn_cast<BoundedRankedTensor>) and then use its convenience methods. In particular this would allow flipping extension attribute to an interface later if needed without changes in users. The implementation details of how stored and actual type remains the same, but C++ API is more shielded from underlying [in particular I do want to be able to reference a context as in value inference above]

Reply:
Nice idea!

Filed https://github.com/openxla/stablehlo/issues/812